### PR TITLE
Script to list transitive dependencies

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

Cocoapods appears to have a different idea of the dependency graph to
SPM which has led to a handful of build failures (e.g.
https://github.com/apple/swift-nio/issues/2073).

This appears to have originated when we dropped the explicit dependency
on `CNIOAtomics` from `NIO` (https://github.com/apple/swift-nio/pull/1719).

We can work around this by listing all transitive dependencies as
requirements in the generated podspecs.

Modifications:

- Add a script to list transitive dependencies for a given module.
- Fix a bug in build_podspecs.sh which did a `pod repo update` even if
  the pods were not being uploaded.
- Update build_podspecs.sh to use transitive dependencies.

Result:

Podspecs should include all transitive dependencies.

### CODE REVIEW

**Comment by glbrntt** [comment]
FWIW I tested the script on HTTP/2 and GRPC since they include package dependencies. I also looked over the generated podspecs.

File: scripts/build_podspecs.sh [hunk 1]
```diff
@@ -75,7 +75,7 @@ for target in "${targets[@]}"; do
 
   while read -r raw_dependency; do
     dependencies+=( "${newline}  s.dependency '$raw_dependency', s.version.to_s" )
-  done < <("${here}/list_topsorted_dependencies.sh" -d "${target#Swift}" | grep -v "NIOPriorityQueue" | sed 's/^NIO/SwiftNIO/')
+  done < <("${here}/list_transitive_dependencies.py" "${target#Swift}" | grep -v "NIOPriorityQueue" | sed 's/^NIO/SwiftNIO/')
```
Comment (weissi) [inline scripts/build_podspecs.sh:78]: should we just drop the `.sh`/`.py` suffix? It doesn't really help and causes issues in situations like this...

Comment (Lukasa) [reply scripts/build_podspecs.sh:78]: Is it causing an issue? We're calling a different script entirely, not the same one rewritten in a different language.

File: scripts/list_transitive_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                        dependency_name, self._packages[package_name], dependencies
+                    )
+
+    def _ensure_package_is_cached(self, package, package_name):
+        if package_name in self._packages:
+            return
+
+        # Find the package dependency with the right name.
+        for package_dependency in package["dependencies"]:
+            if "sourceControl" not in package_dependency:
```
Comment (Lukasa) [inline scripts/list_transitive_dependencies.py:92]: Should we throw an exception in this case rather than ignore it?

**Review by Lukasa — APPROVED** [review]
Basically LGTM except for a couple of notes

File: scripts/build_podspecs.sh (hunk 1 above)
Comment (glbrntt) [reply scripts/build_podspecs.sh:78]: Generally I also have a slightly preference to not having the prefix for scripts _but_ the extension is used by the soundness script when validating the license header so it does serve a small purpose here.

File: scripts/list_transitive_dependencies.py (hunk 2 above)
Comment (glbrntt) [reply scripts/list_transitive_dependencies.py:92]: Yeah, we can probably just remove this check entirely and hard fail if there's no "sourceControl" key.

File: scripts/build_podspecs.sh (hunk 1 above)
Comment (weissi) [reply scripts/build_podspecs.sh:78]: Oh shoot, that's probably also why I named it that in the first place. Not sure if that's a good reason but at least there is a reason :). So yes, consider my point here moot.

**Pull request merged**